webui/src/components/editrecord.jsx

```diff
@@ -550,32 +562,111 @@ const EditRecord = React.createClass({
+            const patch = compare(original, updated);
+            if (!patch || !patch.length) {
+                this.setState({dirty:false});
+                return;
+            }
+            this.applyPatch(patch, 'submitted');
+        }
+        // Revoking the submitted record for review. Going back to the draft mode.
+        if(prevState.revoking !== this.state.revoking && this.state.revoking){
+            updated['publication_state'] = 'draft';
+            const patch = compare(original, updated);
+            this.applyPatch(patch, 'edit');
+        }
+    },
+
     updateRecord(event) {
         event.preventDefault();
         const errors = this.findValidationErrors();
         if (this.state.fileState !== 'done' || pairs(errors).length > 0) {
             this.setState({errors});
             return;
         }
-        const original = this.props.record.get('metadata').toJS();
-        const updated = this.state.record.toJS();
-        const patch = compare(original, updated);
-        if (!patch || !patch.length) {
-            this.setState({dirty:false});
-            return;
-        }
-        const afterPatch = (record) => {
-            if (this.props.isDraft && !this.isForPublication()) {
-                this.props.refreshCache();
-                // TODO(edima): when a draft is publised, clean the state of
-                // records in versioned chain, to trigger a refetch of
-                // versioning data
-                this.setState({dirty:false, waitingForServer: false});
-                notifications.clearAll();
-            } else {
-                browser.gotoRecord(record.id);
+
+        if(this.state.checkboxValue == 'draft'){
+            // Save draft
+            const original = this.props.record.get('metadata').toJS();
+            const updated = this.state.record.toJS();
+            const patch = compare(original, updated);
+            if (!patch || !patch.length) {
+                this.setState({dirty:false});
+                return;
+            }
+            this.applyPatch(patch, 'save_draft');
+        } else if(this.state.checkboxValue == 'published' && this.state.record.get('publication_state') == 'published'){
+            // Editing metadata of a published record (workflow = direct_publish)
+            const original = this.props.record.get('metadata').toJS();
+            const updated = this.state.record.toJS();
+            const patch = compare(original, updated);
+            if (!patch || !patch.length) {
+                this.setState({dirty:false});
+                return;
             }
+            this.applyPatch(patch, 'edit_metadata');
+        } else {
+            // (review workflow) submit a record for review or (direct publish workflow) Publishing a record
+            const record = this.state.record.set('publication_state', this.state.checkboxValue);
+            this.setState({record});
         }
+    },
+
+    applyPatch(patch, caseValue) {
```

Comment:
Good function structure, but the name doesn't show an important part of its function: changing the browser state. I'd rename it to something like `applyPatchAndUpdateBrowser` (this is not a great name, maybe you can find a better one).

Reply:
I'll try ;)